File: matlab/metasignal_mat/tutorials/04_difficulty_dependence.m
%% Tutorial 4 — Difficulty Dependence
% Tests whether each measure changes as task difficulty changes.
% A well-calibrated metacognitive measure should be difficulty-independent.
% Replicates the analysis in Rahnev (2025) Supp Tables 3–5.
%
% Pipeline:
%   1. Simulate subjects with easy / medium / hard difficulty levels
%   2. Compute all 20 measures per difficulty level
%   3. Remove ±3 SD outliers (propagated across levels per subject)
%   4. One-sample t-test: easy − hard

%% Setup
clear; close all; clc

root_dir = fileparts(fileparts(mfilename('fullpath')));
addpath(genpath(fullfile(root_dir, 'helperFunctions')));

variable_names = {'meta-d''', 'AUC2', 'Gamma', 'Phi', 'DeltaConf', ...
    'M-Ratio', 'AUC2-Ratio', 'Gamma-Ratio', 'Phi-Ratio', 'DeltaConf-Ratio', ...
    'M-Diff', 'AUC2-Diff', 'Gamma-Diff', 'Phi-Diff', 'DeltaConf-Diff', ...
    'meta-noise', 'meta-uncertainty', 'd''', 'Criterion', 'Confidence'};

N_MEAS     = 20;
nRatings   = 4;
difficulty = [0.65, 0.75, 0.85];   % hard, medium, easy  (P(correct))
n_levels   = length(difficulty);
n_subjects = 20;
n_per_level = 80;

rng(0);

%% Simulate subjects
% Each subject performs n_per_level trials at each difficulty level.
% Higher accuracy = easier task.

raw = NaN(n_subjects, n_levels, N_MEAS);

for sub = 1:n_subjects
    for lv = 1:n_levels
        acc  = difficulty(lv);
        stim = randi([0,1], n_per_level, 1);
        resp = stim;
        flip = rand(n_per_level,1) > acc;
        resp(flip) = 1 - resp(flip);
        correct = (stim == resp);
        conf = zeros(n_per_level,1);
        conf( correct) = randi([3, nRatings], sum( correct), 1);
        conf(~correct) = randi([1,         2], sum(~correct), 1);
        raw(sub, lv, :) = compute_all_measures(stim, resp, conf, nRatings);
    end
end

fprintf('Simulated %d subjects × %d difficulty levels\n', n_subjects, n_levels);
fprintf('Computed array: %s\n', mat2str(size(raw)));

%% Compute measures per difficulty level

%% 3-SD outlier removal
% Values beyond ±3 SD per measure × level are set to NaN.
% If ANY level is NaN for a given subject × measure, ALL levels are set NaN.

clean = raw;
for m = 1:N_MEAS
    for lv = 1:n_levels
        col  = clean(:, lv, m);
        mu   = nanmean(col);
        sd   = nanstd(col);
        if sd > 0
            clean((col < mu - 3*sd) | (col > mu + 3*sd), lv, m) = NaN;
        end
    end
    % propagate NaN across all levels
    bad = any(isnan(squeeze(clean(:,:,m))), 2);
    clean(bad, :, m) = NaN;
end

removed = sum(isnan(clean(:)) & ~isnan(raw(:)));
fprintf('Values set to NaN by 3-SD removal: %d\n', removed);

%% One-sample t-test: easy − hard
% Index 1 = hard (0.65), index 3 = easy (0.85)

delta = squeeze(clean(:, 3, :)) - squeeze(clean(:, 1, :));  % [sub × meas]

fprintf('\n%-22s  %8s  %10s  %8s  %4s\n', 'Measure', 't', 'p', 'd', 'sig');
fprintf('%s\n', repmat('-', 1, 58));

for m = 1:N_MEAS
    d = delta(:, m);
    d = d(~isnan(d));
    if length(d) < 2, continue; end
    [~, p, ~, stats] = ttest(d);
    t = stats.tstat;
    cd = t / sqrt(length(d));
    if     p < 0.001; stars = '***';
    elseif p < 0.01;  stars = ' **';
    elseif p < 0.05;  stars = '  *';
    else;             stars = '   ';
    end
    fprintf('%-22s  %8.3f  %10.4f  %8.3f  %s\n', variable_names{m}, t, p, cd, stars);
end

%% Visualise difficulty effect
figure('Color','w', 'DefaultAxesFontSize', 9);
level_labels = {'Hard', 'Med', 'Easy'};

for m = 1:N_MEAS
    ax = subplot(4, 5, m);
    col   = squeeze(clean(:, :, m));       % [sub × level]
    means = nanmean(col, 1);
    n_ok  = sum(~isnan(col), 1);
    sems  = nanstd(col, 0, 1) ./ sqrt(max(n_ok, 1));
    errorbar(1:n_levels, means, sems, '-o', 'Color', [0 0.45 0.70], ...
        'LineWidth', 1.2, 'MarkerSize', 4, 'CapSize', 4);
    set(ax, 'XTick', 1:n_levels, 'XTickLabel', level_labels);
    title(variable_names{m}, 'FontSize', 8, 'FontWeight', 'bold');
    yline(0, '--k', 'LineWidth', 0.5);
    box off;
end

% Add a figure-level title using a hidden axes (avoids needing suplabel)
ax_title = axes('Position',[0 0 1 1], 'Visible','off');
text(ax_title, 0.5, 1.01, 'Effect of Difficulty on 20 Metacognitive Measures', ...
    'HorizontalAlignment','center', 'VerticalAlignment','bottom', ...
    'FontSize', 12, 'FontWeight', 'bold', 'Units','normalized');

File: docs/tutorials/live_scripts/04_difficulty_dependence.py
# %% [markdown]
# # Tutorial 4 — Difficulty Dependence
#
# Tests whether each measure changes as task difficulty changes. A good metacognitive
# measure should be difficulty-independent. Replicates Rahnev (2025) Supp Tables 3–5.

# %% [markdown]
# ## Setup

# %%
import numpy as np
import matplotlib.pyplot as plt
from scipy import stats
from metasignal import stdpy

MEASURE_NAMES = [
    "meta-d'", "AUC2", "Gamma", "Phi", "DeltaConf",
    "M-Ratio", "AUC2-Ratio", "Gamma-Ratio", "Phi-Ratio", "DeltaConf-Ratio",
    "M-Diff", "AUC2-Diff", "Gamma-Diff", "Phi-Diff", "DeltaConf-Diff",
    "metaNoise", "metaUncertainty", "d'", "c", "mean_conf",
    "logL", "AIC", "BIC", "AICc", "k", "n",
]
N_MEAS = 26
n_ratings = 4
difficulty = (0.65, 0.75, 0.85)   # hard, medium, easy

def simulate_subject(seed, difficulty_levels=(0.65, 0.75, 0.85), n_per_level=80):
    r = np.random.default_rng(seed)
    trials = []
    for acc in difficulty_levels:
        stim    = r.choice([0, 1], n_per_level)
        resp    = np.where(r.random(n_per_level) < acc, stim, 1 - stim)
        correct = stim == resp
        conf    = np.where(
            correct,
            r.integers(3, n_ratings + 1, n_per_level),
            r.integers(1, 3, n_per_level),
        )
        trials.append((stim, resp, conf))
    return trials

n_subjects = 20
dataset = [simulate_subject(i) for i in range(n_subjects)]
print(f"Simulated {n_subjects} subjects × {len(difficulty)} difficulty levels")

# %% [markdown]
# ## Compute measures per difficulty level

# %%
n_levels = len(difficulty)
raw = np.full((n_subjects, n_levels, N_MEAS), np.nan)

for s_idx, subject_trials in enumerate(dataset):
    for lv_idx, (stim, resp, conf) in enumerate(subject_trials):
        raw[s_idx, lv_idx] = stdpy.compute_all_measures(
            stim, resp, conf, n_ratings=n_ratings
        )

print("Computed array shape:", raw.shape)

# %% [markdown]
# ## 3-SD outlier removal
#
# Values beyond 3 SD per measure/level
# are set to NaN, then propagated across all levels for that subject.

# %%
def remove_3sd_outliers(arr):
    out = arr.copy()
    _, n_lev, n_meas = out.shape
    for m in range(n_meas):
        for lv in range(n_lev):
            col = out[:, lv, m]
            mu, sd = np.nanmean(col), np.nanstd(col, ddof=1)
            if not np.isnan(mu) and sd > 0:
                out[(col < mu - 3*sd) | (col > mu + 3*sd), lv, m] = np.nan
        bad = np.isnan(out[:, :, m]).any(axis=1)
        out[bad, :, m] = np.nan
    return out

clean = remove_3sd_outliers(raw)
removed = int(np.sum(np.isnan(clean) & ~np.isnan(raw)))
print(f"Values set to NaN by 3-SD removal: {removed}")

# %% [markdown]
# ## One-sample t-test: easy − hard

# %%
def ttest_1samp(data):
    x = np.asarray(data, float)
    x = x[~np.isnan(x)]
    n = len(x)
    if n < 2:
        return np.nan, np.nan, np.nan, np.nan
    t, p = stats.ttest_1samp(x, 0)
    return t, n - 1, p, t / np.sqrt(n)

delta = clean[:, 2, :] - clean[:, 0, :]   # easy − hard

print(f"{'Measure':<20} {'t':>8} {'p':>10} {'d':>8} {'sig':>4}")
print("-" * 55)
for m, name in enumerate(MEASURE_NAMES):
    t, df, p, d = ttest_1samp(delta[:, m])
    if np.isnan(t):
        continue
    stars = "***" if p < 0.001 else "**" if p < 0.01 else "*" if p < 0.05 else ""
    print(f"{name:<20} {t:8.3f} {p:10.4f} {d:8.3f} {stars:>4}")

# %% [markdown]
# ## Visualise difficulty effect

# %%
fig, axes = plt.subplots(4, 5, figsize=(16, 10), sharex=True)
axes = axes.flatten()
level_labels = ["Hard", "Med", "Easy"]

for m, (name, ax) in enumerate(zip(MEASURE_NAMES, axes)):
    col = clean[:, :, m]
    means = np.nanmean(col, axis=0)
    n_ok  = np.sum(~np.isnan(col), axis=0)
    sems  = np.nanstd(col, axis=0, ddof=1) / np.sqrt(np.maximum(n_ok, 1))
    ax.errorbar(range(n_levels), means, yerr=sems, marker="o", capsize=4, color="#0072b2")
    ax.set_title(name, fontsize=8, fontweight="bold")
    ax.set_xticks(range(n_levels))
    ax.set_xticklabels(level_labels, fontsize=7)
    ax.axhline(0, color="k", linewidth=0.5, linestyle="--")

plt.suptitle("Effect of Difficulty on 26 Metacognitive Measures", fontsize=12)
plt.tight_layout()
plt.show()
